okhttp-ws/src/main/java/okhttp3/internal/ws/WebSocketWriter.java

```diff
@@ -234,6 +236,20 @@ private void writeMaskedSynchronized(BufferedSource source, long byteCount) thro
 
       buffer.write(source, byteCount);
 
+      if (contentLength != -1) {
+        if (buffer.size() > contentLength) {
+          throw new IOException("Wrote more than reported content length of "
+              + contentLength
+              + " bytes: "
+              + buffer.size());
+        }
+        if (isFirstFrame && buffer.size() > contentLength - 2048 /* segment size */) {
+          // This is either a larger buffered write or a very small write. Defer flushing to the
+          // underlying socket since close() is about to be called and will flush all the content.
+          return;
+        }
+      }
+
       long emitCount = buffer.completeSegmentByteCount();
       if (emitCount > 0) {
```

Comment:
Yeah I had a few iterations before I landed on what's in this PR. I'll rework it so there's no early return.